Mathlib/Tactic/RelCongr/Core.lean

```diff
@@ -0,0 +1,479 @@
+  -- consisting of running `Lean.MVarId.relAssumption` (trying a term together with limited
+  -- forward-reasoning on that term) on each nontrivial hypothesis.
+  let assum g := do
+    let mut hs := #[]
+    -- collect the nontrivial hypotheses
+    for h in ← getLCtx do
+      if !h.isImplementationDetail then
+        hs := hs.push (.fvar h.fvarId)
+    -- run `Lean.MVarId.relAssumption` on each one
+    g.relAssumption hs
+  -- Time to actually run the core tactic `Lean.MVarId.relCongr`!
+  replaceMainGoal (← g.relCongr template names disch assum).2.toList
+
+/-- The `rel` tactic applies "relational congruence" rules to solve a relational goal by
+"substitution".  For example,
+```
+example {a b x c d : ℝ} (h1 : a ≤ b) (h2 : c ≤ d) :
+    x ^ 2 * a + c ≤ x ^ 2 * b + d := by
+  rel [h1, h2]
+```
+In this example we "substitute" the hypotheses `a ≤ b` and `c ≤ d` into the LHS `x ^ 2 * a + c` of
+the goal and obtain the RHS `x ^ 2 * b + d`, thus proving the goal.
+
+The "relational congruence" rules used are the library lemmas which have been tagged with the
+attribute `@[rel_congr]`.  For example, the first example constructs the proof term
+```
+add_le_add (mul_le_mul_of_nonneg_left h1 (pow_bit0_nonneg x 1)) h2
+```
+using the relational congruence lemmas `add_le_add` and `mul_le_mul_of_nonneg_left`.  If there are
+no applicable relational congruence lemmas, the tactic fails.
+
+The tactic attempts to discharge side goals to these "relational congruence" lemmas (such as the
+side goal `0 ≤ x ^ 2` in the above application of `mul_le_mul_of_nonneg_left`) using the tactic
+`rel_congr_discharger`, which wraps `positivity` but can also be extended. If the side goals cannot
+be discharged in this way, the tactic fails. -/
+syntax "rel" " [" term,* "]" : tactic
+
+elab_rules : tactic
+  | `(tactic| rel [$hyps,*]) => do
+    let hyps ← hyps.getElems.mapM (elabTerm · none)
+    let g ← getMainGoal
+    g.withContext do
+    let .app (.app _rel lhs) rhs ← withReducible g.getType'
+      | throwError "rel failed, goal not a relation"
+    unless ← isDefEq (← inferType lhs) (← inferType rhs) do
+      throwError "rel failed, goal not a relation"
+    -- The core tactic `Lean.MVarId.relCongr` will be run with discharger `rel_congr_discharger`
+    let disch g := Term.TermElabM.run' do
+      let [] ← Tactic.run g <| evalTactic (Unhygienic.run `(tactic| rel_congr_discharger))
+        | failure
+    -- The core tactic `Lean.MVarId.relCongr` will be run with "assumption tactic"
+    -- consisting of running `Lean.MVarId.relAssumption` (trying a term together with limited
+    -- forward-reasoning on that term) on each of the listed terms.
+    let assum g := g.relAssumption hyps
+    -- Time to actually run the core tactic `Lean.MVarId.relCongr`!
+    let (_, #[]) ← g.relCongr none [] disch assum
+      | throwError "rel failed, cannot prove goal by 'substituting' the listed relationships"
```

Comment:
This one (`rel`) is strictly weaker than `rel_congr` -- it is actually the version I want for teaching, and in such a context I prefer finishing-only tactics.  I think it's reasonable to discuss whether teaching tactics belong in mathlib (I included it because I thought it might be not just me who wanted to use it for teaching), but if we decide to let it stay, can we preserve my preference on this point?

Reply:
I'm fine with the tactic failing, but I think the error message should show the unsolved goals. I think that is useful even in a teaching tactic, no?

Reply:
Working on it, see question at
https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F-tactics/topic/Alternative.20.22toMessageData.22.20for.20a.20list

Reply:
Done!